File: src/udemy_tutorials_py/udemy_tutorials_py/hw_status_publisher.py
#!/usr/bin/env python3
"""Create a ROS2 node that publishes to a topic using the new custom message
   for the udemy_tutorials_interfaces package.
"""
import sys
import rclpy
from rclpy.node import Node
from udemy_tutorials_interfaces.msg import HardwareStatus


class HardwareStatusPublisher(Node):
    """Create a node that publishes to the hw_status topic using the
       custom HardwareStatus message from the udemy_tutorials_interfaces
    """
    def __init__(self):
        """Intialize the node and create a publisher to the topic."""
        super().__init__("hw_status_publisher")
        self.__publisher = self.create_publisher(
            msg_type=HardwareStatus,
            topic="hw_status",
            qos_profile=10,
        )
        # Create a timer to publish the message every 1 second
        self.create_timer(
            timer_period_sec=1.0,
            callback=self.publish_hw_status,
        )
        # Log a message to indicate the node is running
        self.logger.info("Hardware Status Publisher has been started.")

    @property
    def logger(self):
        """Return the logger for the node."""
        return self.get_logger()

    def publish_hw_status(self):
        """Publish the hardware status message to the topic."""
        hw_status_msg = HardwareStatus()
        hw_status_msg.temperature = 45
        hw_status_msg.are_motors_ready = True
        hw_status_msg.debug_message = "All systems are go!"
        # Publish the message to the topic
        self.__publisher.publish(hw_status_msg)


def main(args=None):
    """Main entry point for the node."""
    rclpy.init(args=args)
    hw_status_publisher = HardwareStatusPublisher()
    rclpy.spin(hw_status_publisher)
    rclpy.shutdown()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/udemy_tutorials_cpp/src/hw_status_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "udemy_tutorials_interfaces/msg/hardware_status.hpp"


class HardwareStatusPublisher : public rclcpp::Node
{
    public:
        HardwareStatusPublisher() : Node("hw_status_publisher")
        {
            publisher_ = this->create_publisher<udemy_tutorials_interfaces::msg::HardwareStatus>(
                "hw_status", 10);
            //This is the constructor of my node
            timer_ = this->create_wall_timer(
                std::chrono::seconds(1),
                //Callback for this timer
                std::bind(&HardwareStatusPublisher::send_hardware_status, this)
            );
            RCLCPP_INFO(this->get_logger(),
                        "HardwareStatusPublisher is ready!");
        } 
    private:
        void send_hardware_status()
        {
            // Create HardwareStatus message to send it
            auto msg = udemy_tutorials_interfaces::msg::HardwareStatus();
            msg.temperature = 35;
            msg.are_motors_ready = false;
            msg.debug_message = "Hello!!!";
            publisher_->publish(msg);
        }
        rclcpp::TimerBase::SharedPtr timer_;
        rclcpp::Publisher<udemy_tutorials_interfaces::msg::HardwareStatus>::SharedPtr publisher_;
};


int main(int argc, char **argv)
{
    //Initialize ROS2 node
    rclcpp::init(argc, argv);
    // Create a HardwareStatusPublisher
    auto node = std::make_shared<HardwareStatusPublisher>();
    // Sping the node
    rclcpp::spin(node);
    // Shut down ROS2 node
    rclcpp::shutdown();
    return 0;
}
